zee/src/editor/mod.rs

```diff
@@ -94,11 +94,16 @@ pub struct Context {
 }
 
 impl Context {
-    pub fn mode_by_filename(&self, filename: impl AsRef<Path>) -> &Mode {
+    pub fn mode_by_filename(&self, filename: impl AsRef<Path>) -> Option<&Mode> {
         self.modes
             .iter()
             .find(|&mode| mode.matches_by_filename(filename.as_ref()))
-            .unwrap_or(&PLAIN_TEXT_MODE)
+    }
+
+    pub fn mode_by_shebang(&self, shebang: &str) -> Option<&Mode> {
+        self.modes
+            .iter()
+            .find(|&mode| mode.matches_by_shebang(shebang))
```

Comment:
If shebang is a variant of `FilenamePattern`, there would be one function and we could continue to return `&Mode` rather than `Option<&Mode>` -- I would like to continue have `Context` own coming up with a `Mode` for any possible file whatsoever and not have a default `PLAIN_TEXT_MODE` potentially duplicated

Reply:
I've been thinking about merging these methods into something like `mode_by_file_pattern` that would always return a `&Mode` but I'm stuck on how to handle the difference in parameters. Two of the variants operate on the filename while the other needs a portion of the file content. To have one method we would have to always pass a portion of the file content to the function in addition to the filename which would require reading at least part of the content of every file.

All of that being said, prior to detecting the mode or creating the buffer `open_file()` is calling `Rope::from_reader()` which I think might be reading the entire file anyway. https://github.com/zee-editor/zee/blob/0b4078470ff6be6cb60388189d08531c368d2cf4/zee/src/editor/mod.rs#L156 I reading that right? If so we already have the Rope created and available in `Buffer::new()` so reading a slice should be quite efficient.

Reply:
Here is an updated version of this code with a limit on the length of text examined for the interpreter directive, an inversion of the order mode checks are performed (filename checks first, then shebang), and a merging of the mode_by_filename() and mode_by_shebang() methods into one, mode_by_file(), which always returns a `&Mode`.

I ended up settling on 256 characters for the maximum length of the shebang directive, matching what linux has done since 2018 (https://lore.kernel.org/lkml/20181112160956.GA28472@redhat.com/) (https://github.com/torvalds/linux/blob/master/include/uapi/linux/binfmts.h)

I haven’t found a satisfactory way of adding a shebang case to FilenamePattern, the biggest stumbling blocks being that the shebang line would need to be passed to the matches() method and the pattern list would need to be sorted/filtered if we wanted to ensure that filename patterns were always examined before shebang directives.